src/Data/Formatter/DateTime.purs

```diff
@@ -0,0 +1,426 @@
+    show ((fromEnum $ T.hour t) `mod` 12) <> cb a
+  Meridiem a →
+    (if (fromEnum $ T.hour t) > 12 then "PM" else "AM") <> cb a
+  Minutes a →
+    show (fromEnum $ T.minute t) <> cb a
+  Seconds a →
+    show (fromEnum $ T.second t) <> cb a
+  Milliseconds a →
+    show (fromEnum $ T.millisecond t) <> cb a
+  Placeholder s a →
+    s <> cb a
+  End → ""
+
+
+format ∷ Formatter → DT.DateTime → String
+format f dt = formatF (flip format dt) dt $ unroll f
+
+formatDateTime ∷ String → DT.DateTime → Either String String
+formatDateTime pattern datetime =
+  parseFormatString pattern <#> flip format datetime
+
+unformat ∷ Formatter → String → Either String DT.DateTime
+unformat f s =
+  let
+    run =
+      runState
+        (P.runParserT (P.PState {input: s, position: initialPos}) $ unformatParser f)
+        initialAccum
+  in
+    case run of
+      Tuple (Left (P.ParseError {message})) state → Left message
+      Tuple _ accum → unformatAccumToDateTime accum
+
+data Meridiem = AM | PM
+
+derive instance eqMeridiem ∷ Eq Meridiem
+
+-- TODO: consider using Map Int
+type UnformatAccum =
+  { year ∷ Maybe Int
+  , month ∷ Maybe Int
+  , day ∷ Maybe Int
+  , hour ∷ Maybe Int
+  , minute ∷ Maybe Int
+  , second ∷ Maybe Int
+  , millisecond ∷ Maybe Int
+  , dayOfWeek ∷ Maybe Int
+  , meridiem ∷ Maybe Meridiem
+  }
+
+initialAccum ∷ UnformatAccum
+initialAccum =
+  { year: Nothing
+  , month: Nothing
+  , day: Nothing
+  , hour: Nothing
+  , minute: Nothing
+  , second: Nothing
+  , millisecond: Nothing
+  , dayOfWeek: Nothing
+  , meridiem: Nothing
+  }
+
+unformatAccumToDateTime ∷ UnformatAccum → Either String DT.DateTime
+unformatAccumToDateTime a =
+  DT.DateTime
+    <$> (D.canonicalDate
+           <$> (maybe (Left "incorrect year") pure $ toEnum $ fromMaybe zero a.year)
+           <*> (maybe (Left "incorrect month") pure $ toEnum $ fromMaybe one a.month)
+           <*> (maybe (Left "incorrect day") pure $ toEnum $ fromMaybe one a.day))
+    <*> (T.Time
+           <$> (maybe
+                  (Left "incorrect hour") pure
+                  $ toEnum
+                  $ (if a.meridiem == Just PM then (_ + 12) else id)
+                  $ fromMaybe zero a.hour)
+           <*> (maybe (Left "incorrect minute") pure $ toEnum $ fromMaybe zero a.minute)
+           <*> (maybe (Left "incorrect second") pure $ toEnum $ fromMaybe zero a.second)
+           <*> (maybe (Left "incorrect millisecond") pure $ toEnum $ fromMaybe zero a.millisecond))
+
+
+unformatFParser
+  ∷ ∀ a
+  . (a → P.ParserT String (State UnformatAccum) Unit)
+  → FormatterF a
+  → P.ParserT String (State UnformatAccum) Unit
+unformatFParser cb = case _ of
+  YearFull a → do
+    ds ← some digit
+    when (Arr.length ds /= 4) $ P.fail "Incorrect full year"
+    lift $ modify _{year = Just $ foldDigits ds}
+    cb a
+  YearTwoDigits a → do
+    ds ← some digit
+    when (Arr.length ds /= 2) $ P.fail "Incorrect 2digit year"
+    lift $ modify _{year = Just $ foldDigits ds + 1900}

```

Comment:
This case is even trickier than the formatting case. I'd expect the 2-digit year `16` to be `2016` rather than `1916`, so we should probably instate some switch where numbers over a certain value go into the 1900s, and numbers below go into the 2000s. How about 1970? Since that goes with the unix epoch also. So `70 -> 1970`, `69 -> 2069`, etc.

We could probably move it even nearer the current day, but the choice is pretty arbitrary.